src/front/wgsl/mod.rs

```diff
@@ -2045,104 +2378,222 @@ impl Parser {
             self.scopes.last().map(|&(ref scope, _)| scope.clone()),
             Some(Scope::PrimaryExpr)
         );
-        let ty_resolution = match self.lookup_type.get(type_name) {
-            Some(&handle) => TypeResolution::Handle(handle),
-            None => match self.parse_type_decl_impl(
-                lexer,
-                TypeAttributes::default(),
-                type_name,
-                ctx.types,
-                ctx.constants,
-            )? {
-                Some(inner) => TypeResolution::Value(inner),
-                None => return Ok(None),
-            },
+        let dst_ty = match self.lookup_type.get(type_name) {
+            Some(&handle) => ConstructorType::Struct(handle),
+            None => {
+                match self.parse_constructor_type(lexer, type_name, ctx.types, ctx.constants)? {
+                    Some(inner) => inner,
+                    None => {
+                        match self.parse_type_decl_impl(
+                            lexer,
+                            TypeAttributes::default(),
+                            type_name,
+                            ctx.types,
+                            ctx.constants,
+                        )? {
+                            Some(_) => {
+                                let span = self.pop_scope(lexer);
```

Comment:
why are we popping the scope here? I don't see where the corresponding push is

Reply:
I'll use `self.current_byte_offset()` to get the span instead. I was thinking since it will error out it wouldn't matter.

Reply:
Actually, to use `current_byte_offset` there I would need the start of the span as well. The start of the span would have to be passed to the `parse_construction` function since it already gets the parsed type name from its caller.

Should I pass the span of the type into the `parse_construction` fn or leave it as is? Does `pop_scope` cause issues in this instance? (I'm not too familiar with that part of the code)

Reply:
Just keep the push and pop corresponding to each other. Having this `pop` without the relevant `push` is the weird part. If you need to pass the span into `parse_construction` - so be it.

Reply:
Ah, I think I get what your concern is now.

The `parse_construction` expects there to be a `PrimaryExpr` scope pushed already.

https://github.com/gfx-rs/naga/blob/012f2a6b2e9328e93939429705fdf9102d9f69b7/src/front/wgsl/mod.rs#L2047

Nonetheless, I changed the fn signature to pass the span trough it. It's a more accurate span actually (doesn't point to whitespace).